src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs

```diff
@@ -701,20 +707,21 @@ private void CheckForHttp2ConnectionInjection()
         {
             Debug.Assert(HasSyncObjLock);
 
-            if (!_http2RequestQueue.TryPeekUncanceledRequest(this, out HttpRequestMessage? request))
-            {
-                return;
-            }
+            _http2RequestQueue.PruneCompletedRequestsFromHeadOfQueue(this);
 
             // Determine if we can and should add a new connection to the pool.
             if ((_availableHttp2Connections?.Count ?? 0) == 0 &&                            // No available connections
                 !_pendingHttp2Connection &&                                                 // Only allow one pending HTTP2 connection at a time
-                (_associatedHttp2ConnectionCount == 0 || EnableMultipleHttp2Connections))   // We allow multiple connections, or don't have a connection currently
+                _http2RequestQueue.Count > 0 &&                                             // There are requests left on the queue
```

Comment:
Nit: just curious why in some cases we're using > 0 and in some != 0... surely Count won't be negative?

Reply:
No real reason, behavior should be idential. I'll change it to match the `> 0` we're already using.